client/qubesdb-cmd.c

```diff
@@ -5,79 +5,135 @@
+#endif
+            if (fwrite(val, 1, i, stdout) != i) {
+                perror("fwrite");
+                QUBES_SECURE_ZERO(old_val, len + 1);
+                return 1;
+            }
+#ifndef WIN32
+        } else do {
+            ssize_t written = write(1, val, i);
+            if (written == -1) {
+                int e = errno;
+                if (e != EAGAIN && e != EWOULDBLOCK && e != EINTR) {
+                    perror("write");
+                    QUBES_SECURE_ZERO(old_val, len + 1);
+                    return 1;
+                }
+            } else {
+                val += (size_t)written;
+                i -= (size_t)written;
+            }
+        } while (i > 0);
+#endif
+        QUBES_SECURE_ZERO(old_val, len + 1);
+        return 0;
     }
-    len = strlen(val);
     for (i = 0; i < len; i++) {
+        int status;
         if (val[i] >= 0x20 && (unsigned char)val[i] < 0x80)
-            printf("%c", val[i]);
-        else
+            status = printf("%c", val[i]);
+        else {
 #ifndef WIN32
-            printf("\\x%02hhx", val[i]);
+            status = printf("\\x%02hhx", val[i]);
 #else
             /* windows doesn't support 'h' modifier */
-            printf("\\x%02x", val[i]);
+            status = printf("\\x%02x", val[i]);
 #endif
+        }
+        if (status < 0) {
+            perror("printf");
+            return 1;
+        }
     }
+    if (fputs("\n", stdout) == EOF) {
+        perror("fputs");
+        return 1;
+    }
+    return 0;
 }
 
-int cmd_read(qdb_handle_t h, int argc, char **args) {
+static int cmd_read(qdb_handle_t h, int argc, char **args) {
     int i;
-    char *value;
-    int anything_failed = 0;
+    char *value, *path;
+    int anything_failed = 0, is_enoent = 0;
 
     for (i=0; i < argc; i++) {
         value = qdb_read(h, args[i], NULL);
-        if (opt_wait) {
-            qdb_watch(h, args[i]);
-            while (!(value = qdb_read(h, args[i], NULL)))
-            {
-                qdb_read_watch(h);
+        if (!opt_wait)
+            is_enoent = !value && errno == ENOENT;
+        else if (!value) {
+            anything_failed |= qdb_watch(h, args[i]) != 1;
+            while (!(value = qdb_read(h, args[i], NULL))) {
+                if ((path = qdb_read_watch(h))) {
+                    free(path);
+                } else {
+                    anything_failed = 1;
+                }
             }
         }
         if (value) {
             if (opt_fullpath)
                 printf("%s = ", args[i]);
-            encode_and_print_value(value);
-            printf("\n");
+            anything_failed |= encode_and_print_value(value, opt_fullpath);
         } else {
             if (!opt_quiet) {
                 fprintf(stderr, "Failed to read %s\n", args[i]);
             }
-            anything_failed = 1;
+            if (!is_enoent)
+                anything_failed = 1;
+        }
+        if (opt_rm) {
+            if (qdb_rm(h, args[i]) != 1)
```

Comment:
What's the advantage to move this logic here instead of letting the caller use qubesdb-rm or similar?

Reply:
It simplifies the random seed service, and it was trivial to implement.